New exercise assignment based on recognition/recall

merleschoen opened the pull request:
This pull request introduces additional difficulty categories based on the memory retrieval concepts of recognition and recall. Research from my thesis highlights the need for more nuanced difficulty levels.

**New Difficulty Levels:** I've added new difficulty categories for exercises, focusing on recognition and recall. Bookmarks are initially tested for receptive/productive recognition, and will be tested for receptive/productive recall, once they have at least 2 consecutive correct answers for recognition and a cooling interval of 2 or more.

![image](https://github.com/user-attachments/assets/a12ce936-0fd4-4bfa-88ba-4c1f777ccf94)

![image](https://github.com/user-attachments/assets/cab474aa-41a0-4b11-b67b-4506d6cb37be)

**API Integration:** This PR relies on changes from the `consecutive_correct-for-recognition/recall-categories` API branch, which includes `consecutive_correct_answers` data now being passed to the frontend.

**Component Refactoring:** The `exerciseSequenceTypes` component has been refactored by @tfnribeiro

**Notes:**
For the new category, I've used `cognitiveFocus` for now. I've also considered `memoryRetrieval`. If you have any suggestions or preferences, please let me know!

tfnribeiro commented:
I just merged the latest branch from Master into this so the diff doesn't show changes already in Master.

mircealungu commented:
The general term for Recognize/Recall is hard to define. Your's is good, and `memoryTask`, or `cognitiveTask` might also work. The question is what are we talking about: the task that is being tested by the exercise, or is it the focus of the exercise. Although it's hard to argue that this is exactly the only focus of the exercise... ?

mircealungu reviewed (approved):
overall looks good. 
small suggestions in the comments.

merleschoen commented:
Thanks for the feedback! I've implemented all the changes you have suggested and changed the name to `memoryTask` - I think that's a better name for it.

mircealungu commented on src/exercises/exerciseSequenceTypes.js:
```diff
@@ -17,12 +17,14 @@ const EX_Match = {
-  requiredBookmarks: 3,
+  requiredBookmarks: 1,
   learningCycle: "productive",
+  memoryTask: "recognition",
```
should this not be `MEMORY_TASK.RECOGNITION`?

mircealungu reviewed (changes requested):
Small request for replacing all the strings with the frozen object :)

mircealungu commented on src/exercises/exerciseSequenceTypes.js:
```diff
@@ -141,3 +152,43 @@ export {
   NUMBER_OF_BOOKMARKS_TO_PRACTICE,
 };
+
+// Function to validate that every combination of learningCycle and memoryTask has at least one exercise with requiredBookmarks = 1
```
super fancy :)
